### PR TITLE
Bluetooth: Host: Fix unsafe ccc cast

### PR DESCRIPTION
**Pull request by HaavardRei**
Contains two commits:

- Refactors `bt_gatt_is_subscribed` when checking the characteristics declaration to avoid undefined behavior.
- Fixes an unsafe cast in the same function when extracting the CCC attribute value.

### CODE REVIEW

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -3523,16 +3522,23 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
...
-	ccc = attr->user_data;
+	if (!attr->read) {
+		LOG_ERR("Read method not set");
+		return false;
+	}
 
-	/* Check if the connection is subscribed */
-	for (size_t i = 0; i < BT_GATT_CCC_MAX; i++) {
-		const struct bt_gatt_ccc_cfg *cfg = &ccc->cfg[i];
+	len = attr->read(conn, attr, &ccc_bits, sizeof(ccc_bits), 0);
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/gatt.c:3530]: To treat it as an integer, we must parse the read result (as a little-endian integer). Something like the following:

```suggestion
	len = attr->read(conn, attr, ccc_bits_encoded, sizeof(ccc_bits), 0);
	ccc_bits = sys_get_le16(ccc_bits_encoded);
	
	// Declaration should be above. I put it here for a nicer diff.
	uint8_t ccc_bits_encoded[sizeof(ccc_bits)];
```

Feel free to experiment with helper functions here, if you want.

File: subsys/bluetooth/host/gatt.c [hunk 2]
```diff
@@ -3523,16 +3522,23 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
...
+	if (len < 0) {
+		LOG_ERR("Failed to read attribute (err %zd)", len);
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/gatt.c:3532]: I think we should print the pointer to the attribute. It could be useful for debugging.

```suggestion
		LOG_ERR("Failed to read attribute %p (err %zd)", attr, len);
```

**Review by alwa-nordic — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/gatt.c (hunk 1 above)
Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c]: Added, tested before and after using the ccc_store bsim test: https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bsim/bluetooth/host/gatt/ccc_store/src/peripheral.c#L147, both works but your suggestion is more correct I think :)

File: subsys/bluetooth/host/gatt.c [hunk 3]
```diff
@@ -3464,7 +3464,10 @@ static uint8_t disconnected_cb(const struct bt_gatt_attr *attr, uint16_t handle,
 bool bt_gatt_is_subscribed(struct bt_conn *conn,
 			   const struct bt_gatt_attr *attr, uint16_t ccc_type)
 {
-	const struct _bt_gatt_ccc *ccc;
+	uint16_t ccc_bits;
+	uint8_t properties;
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:3468]: `uint8_t properties;` can be move to the `if (!bt_uuid_cmp(attr->uuid, BT_UUID_GATT_CHRC)) {` clause to reduce scope of the variable

File: subsys/bluetooth/host/gatt.c [hunk 4]
```diff
@@ -3523,16 +3523,25 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
...
-	ccc = attr->user_data;
+	if (!attr->read) {
+		LOG_ERR("Read method not set");
+		return false;
+	}
 
-	/* Check if the connection is subscribed */
-	for (size_t i = 0; i < BT_GATT_CCC_MAX; i++) {
-		const struct bt_gatt_ccc_cfg *cfg = &ccc->cfg[i];
+	len = attr->read(conn, attr, ccc_bits_encoded, sizeof(ccc_bits), 0);
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:3531]:
```suggestion
	len = attr->read(conn, attr, ccc_bits_encoded, sizeof(ccc_bits_encoded), 0);
```

File: subsys/bluetooth/host/gatt.c [hunk 5]
```diff
@@ -3523,16 +3523,25 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
...
+	if (len < 0) {
+		LOG_ERR("Failed to read attribute %p (err %zd)", attr, len);
+		return false;
+	} else if (len != sizeof(ccc_bits)) {
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:3535]:
```suggestion
	} else if (len != sizeof(ccc_bits_encoded)) {
```

File: subsys/bluetooth/host/gatt.c [hunk 6]
```diff
@@ -3464,7 +3464,10 @@ static uint8_t disconnected_cb(const struct bt_gatt_attr *attr, uint16_t handle,
 bool bt_gatt_is_subscribed(struct bt_conn *conn,
 			   const struct bt_gatt_attr *attr, uint16_t ccc_type)
 {
-	const struct _bt_gatt_ccc *ccc;
+	uint16_t ccc_bits;
+	uint8_t properties;
+	uint8_t ccc_bits_encoded[sizeof(ccc_bits)];
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:3469]: This could in theory also be
```suggestion
	uint16_t ccc_bits_encoded;
```

where you then can do `ccc_bits = sys_le16_to_cpu(ccc_bits_encoded)`

Either way works though, but this would make it more similar to how `properties` is handled

Comment (alwa-nordic) [reply subsys/bluetooth/host/gatt.c:3468]: I prefer the array for a practical reason and a simplicity (aesthetic) reason:
 - The array cannot be confused with the parsed value.
 - `sys_le16_to_cpu` is not a nice mathematical function. It takes in a number, and what number comes out depends on the CPU architecture. This is weird. `sys_get_le16` on the other hand always does the same thing.
 - `uint16_t ccc_bits_encoded` has the same problem as above. It will have different values on different platforms for the same meaning.

Comment (Thalley) [reply subsys/bluetooth/host/gatt.c:3468]: Based on that, are you still OK with `properties` not being an array then? Like `uint8_t properties[1]`? 

> sys_le16_to_cpu is not a nice mathematical function. It takes in a number, and what number comes out depends on the CPU architecture. This is weird. sys_get_le16 on the other hand always does the same thing.

`sys_get_le16` works the same way as `sys_le16_to_cpu`; take a little-endian value and convert it to host endianess. I don't understand why you prefer one over the other. `sys_le16_to_cpu` can possibly be considered slightly better in terms of performance, as it is a no-op on little-endian archs, where `sys_get_le16` always does the operation.

Comment (Thalley) [reply subsys/bluetooth/host/gatt.c:3468]: ^_^

Not strong preferences from me on this topic, though I am generally in favor of consistency :) Fine to keep as is

File: subsys/bluetooth/host/gatt.c [hunk 7]
```diff
@@ -3523,16 +3524,25 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
...
-	for (size_t i = 0; i < BT_GATT_CCC_MAX; i++) {
-		const struct bt_gatt_ccc_cfg *cfg = &ccc->cfg[i];
+	ccc_bits = sys_get_le16(ccc_bits_encoded);
 
-		if (bt_conn_is_peer_addr_le(conn, cfg->id, &cfg->peer) &&
-		    (ccc_type & ccc->cfg[i].value)) {
-			return true;
-		}
+	/* Check if the CCC bits match the subscription type */
+	if (ccc_bits & ccc_type) {
```
[the remark this answers is not in the document]
Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c:3544]: The read function for CCCs (included in the ccc declaration here: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/bluetooth/gatt.h#L1045), leads to this check: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/gatt.c#L2159-L2160, so the check is still included!

(Called here in read: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/gatt.c#L2179)

[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/host/gatt.c:3544]: This is assuming that the `attr->read` is `bt_gatt_attr_read_ccc`, isn't it? In your previous PR you removed the assumption that the GATT macros were used, but here you seem to rely on `BT_GATT_CCC_MANAGED` being used. If an application implement its own version of `bt_gatt_attr_read_ccc` that does not have the `bt_conn_is_peer_addr_le(conn, cfg->id, &cfg->peer)` check, then this is changing behavior. 

Based on the reasoning from https://github.com/zephyrproject-rtos/zephyr/pull/80570 I think it is OK, but we are expecting that users either use the GATT macros (which contradicts the need for https://github.com/zephyrproject-rtos/zephyr/pull/80570) or that the users uses `bt_gatt_attr_read_ccc` or implements their own function that ensure this check. None of that is documented, and if anyone relied on the check here instead of having it in their own read function, then their software may start to silently break. 

TL;DR: It is OK to not have the check here, but we should probably mention it in the release notes or migration guides, as this is changing the behavior of this function based on undocumented expectations.

[the remark this answers is not in the document]
Comment (alwa-nordic) [reply subsys/bluetooth/host/gatt.c:3544]: > This is assuming that the `attr->read` is `bt_gatt_attr_read_ccc`, isn't it?

No, it does not care whether `attr->read` is `bt_gatt_attr_read_ccc`.

All we care about is that `attr->uuid` is `BT_UUID_GATT_CCC`. Then `attr->read` must act according to spec and give the value of the CCC.

In GATT, a remote client can see if it's subscribed by reading the CCC value with a normal ATT read. So to see whether a remote client is subscribed, we just emulate a remote read using `attr->read(conn)`, and we immediately get the answer we want.

[the remark this answers is not in the document]
Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c:3544]: Fair point!
I added a section about this to the migration guide, though I am a bit uncertain on how much we should recommend adding this check. I went for "this check should be included", but if we want a weaker or stronger recommendation I can change it :)

[the remark this answers is not in the document]
Comment (alwa-nordic) [reply subsys/bluetooth/host/gatt.c:3544]: There is no fair point. Please think carefully about what the "`bt_conn_is_peer_addr_le(conn, cfg->id, &cfg->peer)` check" actually is.

[the remark this answers is not in the document]
Comment (alwa-nordic) [reply subsys/bluetooth/host/gatt.c:3544]: Answer: It part of the the lookup of `cfg`. It's meaningless by itself. It's not a "check" at all.

```c
	for (size_t i = 0; i < BT_GATT_CCC_MAX; i++) {
		const struct bt_gatt_ccc_cfg *cfg = &ccc->cfg[i];

		if (bt_conn_is_peer_addr_le(conn, cfg->id, &cfg->peer)) {
			/* Found the `cfg` for `conn`. */
		}
	}
```

[the remark this answers is not in the document]
Comment (Thalley) [reply subsys/bluetooth/host/gatt.c:3544]: Fair point @alwa-nordic 

GATT is basically underdocumented in how it works and how applications using their own functions should work, so when changing anything here it can break stuff that may have worked previously due to assumptions from applications. 

This PR is fine with or without the migration guide note, and the real issue is the lack of documentation IMO :)

**Review by Thalley — DISMISSED** [review]

**Review by Thalley — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Comment by HaavardRei** [comment]
@jhedberg Should we add this to the 4.0 milestone?

**Comment by dkalowsk** [comment]
I do not see a GitHub issue attached, skipping on this merge round for 4.0
